README.md
# CoinTools

This is a collection of Ruby scripts and library classes that let you check cryptocurrency prices on various services.

[![Gem Version](https://badge.fury.io/rb/cointools.svg)](https://badge.fury.io/rb/cointools) [![Build Status](https://travis-ci.org/mackuba/cointools.svg?branch=master)](https://travis-ci.org/mackuba/cointools)


## Installation

CoinTools requires Ruby version 2.3 or newer.

To use the scripts from the command line, install the gem (depending on your configuration, you might need to add `sudo`):

```
gem install cointools
```

To use the code as a library, add it to your Gemfile:

```ruby
gem 'cointools'
```

## Cryptowatch

[Cryptowat.ch](https://cryptowat.ch) is a coin charts site owned by Kraken which tracks historical cryptocurrency prices directly on many exchanges. It provides data from specific markets on specific exchanges, but no general rankings or average market prices.

To check past price of a given coin on a chosen exchange, pass the exchange and market name and a properly formatted timestamp:

```
cryptowatch bitfinex btcusd "2017-12-17 13:00"
```

To check the current price, skip the timestamp:

```
cryptowatch bitfinex btcusd
```

By default the price is printed in a verbose format that includes the requested market name and the time of the returned price. Add a `-q/--quiet` option to only print the price as a single number (e.g. to pass it further to another script).

You can fetch a list of available exchanges and markets using these commands:

```
cryptowatch --exchanges
cryptowatch --markets bithumb
```

In code:

```ruby
require 'cointools'  # or 'cointools/cryptowatch'
cryptowatch = CoinTools::Cryptowatch.new

exchange = cryptowatch.exchanges.first

list = cryptowatch.get_markets(exchange)
market = list.select { |x| x =~ /ltc/ && x !~ /btc/ }.first.upcase

result = cryptowatch.get_price(exchange, market, Time.now - 86400)
puts "#{market} yesterday: #{result.price}"

result = cryptowatch.get_current_price(exchange, market)
puts "#{market} today: #{result.price}"
```

The result object contains the requested price and (for historical prices) the actual timestamp of the found price, which might slightly differ from the timestamp passed in the argument (the earlier the date, the less precise the result).

The API is rate limited to 8 seconds of CPU time per hour - you can check the `api_time_spent` and `api_time_remaining` properties of the result object to see how much time allowance you have left.

If you need to do a large amount of lookups in a short period of time, try the alternative `#get_price_fast` method (or `-f` option on the command line). This method tries to guess which data set is the most appropriate for a given point in the past and requests only that set instead of all of them, which should use significantly less API time allowance. This however relies on an undocumented API behavior, so it's not guaranteed to return data and keep working in the future.


## CoinMarketCap

[CoinMarketCap](https://coinmarketcap.com) is by far the most popular site for checking coin rankings and BTC/USD prices for all coins available on the market. The API however only returns current average coin prices.

To look up a coin's price, you need to pass its name as used on CoinMarketCap:

```
cmcap power-ledger
```

Alternatively, you can pass the cryptocurrency's symbol using the `-s` or `--symbol` parameter:

```
cmcap -s powr
```

However, be aware that coin symbols are not unique on CoinMarketCap - there are at least a few examples of duplicates. In such case, the first coin that matches (probably the older one) will be returned.

By default the price is printed in a verbose format that includes the requested coin symbol and the time of the returned price. Add a `-q/--quiet` option to only print the price as a single number (e.g. to pass it further to another script).

You can also use the `-b` or `--btc-price` flag to request a price in BTC instead of USD:

```
cmcap power-ledger -b
```

Or you can request the price in one of the ~30 other supported fiat currencies with `-c` or `--convert-to`:

```
cmcap request-network -cEUR
```

You can print a list of supported fiat currencies with `cmcap --fiat-currencies`.

Same things in code:

```ruby
require 'cointools'  # or 'cointools/coinmarketcap'
cmc = CoinTools::CoinMarketCap.new

p CoinTools::CoinMarketCap::FIAT_CURRENCIES

ltc = cmc.get_price('litecoin')
puts "LTC: #{ltc.usd_price} USD / #{ltc.btc_price} BTC"

xmr = cmc.get_price_by_symbol('xmr')
puts "XMR: #{xmr.usd_price} USD / #{xmr.btc_price} BTC"

eth = cmc.get_price('ethereum', convert_to: 'EUR')
puts "ETH: #{eth.converted_price} EUR"
```

The soft rate limit for the API is 30 requests per minute (for API v2).

Note: since in the v2 API specific coin tickers can only be looked up using CoinMarketCap's internal numeric ids (e.g. Ethereum = 1027), both lookup methods available here - by the coin name ("slug") and symbol - have to first download a `/listings` JSON with a mapping of all coins on the site. The result of that call is cached in a `CoinTools::CoinMarketCap` object, so if you do many lookups in code, in one go or over some period of time, it's recommended to reuse the object instead of recreating it each time.

The API only allows one additional currency apart from USD, so if you pass a `convert_to` parameter, `btc_price` will not be returned, only `converted_price`.


### Getting all coin prices

You can also download the whole table of 1600+ coins listed on CoinMarketCap using the `get_all_prices` method. The method also takes a `convert_to` parameter, which works as described above. This method has to use paging to download data in batches and currently makes around 16 calls (which takes about half a minute), so do not call it repeatedly or you might go over the rate limit.

You can pass a block to this method to receive each batch of coins as they arrive. Take into account however that coins are not sorted by rank, but by numeric id (or roughly by creation date), so if you want e.g. top 100 coins, you still need to download the whole list. This is mostly intended to let you track progress in some way, e.g.:

```ruby
require 'cointools/coinmarketcap'
cmc = CoinTools::CoinMarketCap.new

coins = cmc.get_all_prices { |l| print '.' }
puts

coins.first(50).each do |c|
  puts c.symbol.ljust(5) + c.market_cap.to_i.to_s.rjust(15)
end
```


## CoinCap

[CoinCap.io](https://coincap.io) is a site similar to CoinMarketCap with an API that includes historical coin prices (with decreasing precision the further into the past you look).

To check past price of a given coin, pass the coin's symbol and a properly formatted timestamp:

```
coincap xmr "2018-04-01 13:00"
```

To check the current price, skip the timestamp:

```
coincap xmr
```

By default the price is printed in a verbose format that includes the requested coin symbol and the time of the returned price. Add a `-q/--quiet` option to only print the price as a single number (e.g. to pass it further to another script).

You can also use the `-b` or `--btc-price` flag to request a price in BTC instead of USD, or `-e` or `--eur-price` for EUR:

```
coincap xmr -b
coincap xmr -e
```

These are however only supported for current prices - past prices are only listed in USD.

In code:

```ruby
require 'cointools'  # or 'cointools/coincap'
coincap = CoinTools::CoinCap.new

result = coincap.get_price('XMR', Time.now - 86400)
puts "XMR yesterday: $#{result.usd_price} (#{result.time})"

result = coincap.get_current_price('XMR')
puts "XMR today: $#{result.usd_price} / €#{result.eur_price} / ₿#{result.btc_price}"
```

The result object contains the requested price and (for historical prices) the actual timestamp of the found price, which might slightly differ from the timestamp passed in the argument (the earlier the date, the less precise the result).

At the moment there don't seem to be any official rate limits for the API.


## Credits & contributing

Copyright © 2018 [Kuba Suder](https://mackuba.eu). Licensed under [MIT License](http://opensource.org/licenses/MIT).

If you'd like to help me extend the scripts with some additional features or add support for new services, [send me a pull request](https://github.com/mackuba/cointools/pulls).


CHANGELOG.md
#### Version 0.5.1 (19.07.2018)

* fixed bug in `CoinMarketCap#get_all_prices` that caused duplicated entries

#### Version 0.5.0 (12.07.2018)

* switched to CoinMarketCap API v2, CMC's `get_price` and `get_price_by_symbol` calls now return data like coin name/symbol, rank and market cap
* new `get_all_prices` method in CMC that returns all 1600+ coins sorted by rank
* renamed `--list-exchanges` to `--exchanges`, `--list-markets` to `--markets`, `--list-fiat-currencies` to `--fiat-currencies`, `--fiat-currency` to `--convert-to`
* all commands print verbose output by default, use `--quiet/-q` to print only the price
* time can be passed as a string to `get_price` methods
* improved error handling
* added a huge amount of tests

#### Version 0.4.0 (23.05.2018)

* added CoinCap.io class and `coincap` command
* renamed CoinMarketCap command to `cmcap`
* removed BitBay class and command - BitBay is now supported on Cryptowatch (including past prices since March)
* fixed CoinMarketCap calls with currency conversion
* added "fast" method for Cryptowat.ch
* include API time allowance in Cryptowat.ch responses

#### Version 0.3.0 (26.03.2018)

* added BitBay class and `bitbay-price` command

#### Version 0.2.1 (6.03.2018)

* changed license to MIT

#### Version 0.2.0 (29.01.2018)

* added CoinMarketCap class & `coinmcap` binary

#### Version 0.1.2 (24.01.2018)

* fixed `--help` command

#### Version 0.1.1 (24.01.2018)

* first working release


lib/cointools/cryptowatch.rb
require_relative 'base_struct'
require_relative 'errors'
require_relative 'request'
require_relative 'utils'

require 'net/http'
require 'uri'

module CoinTools
  class Cryptowatch
    BASE_URL = "https://api.cryptowat.ch"

    # we expect this many days worth of data for a given period precision (in seconds); NOT guaranteed by the API
    DAYS_FOR_PERIODS = {
      60 => 3, 180 => 10, 300 => 15, 900 => 2 * 30, 1800 => 4 * 30, 3600 => 8 * 30,
      7200 => 365, 14400 => 1.5 * 365, 21600 => 2 * 365, 43200 => 3 * 365, 86400 => 4 * 365
    }

    DataPoint = BaseStruct.make(:price, :time, :api_time_spent, :api_time_remaining)


    def exchanges
      @exchanges ||= get_exchanges
    end

    def get_markets(exchange)
      raise InvalidExchangeError if exchange.to_s.empty?

      url = URI("#{BASE_URL}/markets/#{exchange}")

      response = Request.get(url)

      case response
      when Net::HTTPSuccess
        json = Utils.parse_json(response.body)
        raise JSONError.new(response) unless json.is_a?(Hash) && json['result'].is_a?(Array)

        return json['result'].select { |m| m['active'] == true }.map { |m| m['pair'] }.sort
      when Net::HTTPNotFound
        raise UnknownExchangeError.new(response)
      when Net::HTTPClientError
        raise BadRequestError.new(response)
      else
        raise ServiceUnavailableError.new(response)
      end
    end

    def get_price(exchange, market, time = nil)
      raise InvalidExchangeError if exchange.to_s.empty?
      raise InvalidSymbolError if market.to_s.empty?

      if time.nil?
        return get_current_price(exchange, market)
      elsif time.is_a?(String)
        time = Utils.parse_time(time)
      end

      (time <= Time.now) or raise InvalidDateError.new('Future date was passed')
      (time.year >= 2009) or raise InvalidDateError.new('Too early date was passed')

      unixtime = time.to_i
      current_time = Time.now.to_i
      url = URI("#{BASE_URL}/markets/#{exchange}/#{market}/ohlc?after=#{unixtime}")

      response = Request.get(url)

      case response
      when Net::HTTPSuccess
        json = Utils.parse_json(response.body)
        raise JSONError.new(response) unless json.is_a?(Hash)

        data = json['result']
        allowance = json['allowance']
        raise JSONError.new(response) unless data.is_a?(Hash) && allowance.is_a?(Hash)

        timestamp, o, h, l, c, volume = best_matching_record(data, unixtime, current_time)
        raise NoDataError.new(response, 'No price data returned') unless timestamp && o

        actual_time = Time.at(timestamp)

        return DataPoint.new(
          price: o,
          time: actual_time,
          api_time_spent: allowance['cost'],
          api_time_remaining: allowance['remaining']
        )
      when Net::HTTPNotFound
        raise UnknownCoinError.new(response)
      when Net::HTTPClientError
        raise BadRequestError.new(response)
      else
        raise ServiceUnavailableError.new(response)
      end
    end

    def get_current_price(exchange, market)
      raise InvalidExchangeError if exchange.to_s.empty?
      raise InvalidSymbolError if market.to_s.empty?

      url = URI("#{BASE_URL}/markets/#{exchange}/#{market}/price")

      response = Request.get(url)

      case response
      when Net::HTTPSuccess
        json = Utils.parse_json(response.body)
        raise JSONError.new(response) unless json.is_a?(Hash)

        data = json['result']
        allowance = json['allowance']
        raise JSONError.new(response) unless data.is_a?(Hash) && allowance.is_a?(Hash)

        price = data['price']
        raise NoDataError.new(response) unless price

        return DataPoint.new(
          price: price,
          time: nil,
          api_time_spent: allowance['cost'],
          api_time_remaining: allowance['remaining']
        )
      when Net::HTTPNotFound
        raise UnknownCoinError.new(response)
      when Net::HTTPClientError
        raise BadRequestError.new(response)
      else
        raise ServiceUnavailableError.new(response)
      end
    end

    def get_price_fast(exchange, market, time = nil)
      raise InvalidExchangeError if exchange.to_s.empty?
      raise InvalidSymbolError if market.to_s.empty?

      if time.nil?
        return get_current_price(exchange, market)
      elsif time.is_a?(String)
        time = Utils.parse_time(time)
      end

      (time <= Time.now) or raise InvalidDateError.new('Future date was passed')
      (time.year >= 2009) or raise InvalidDateError.new('Too early date was passed')

      period = precision_for_time(time)

      if period.nil?
        return get_price(exchange, market, time)
      end

      unixtime = time.to_i
      current_time = Time.now.to_i
      url = URI("#{BASE_URL}/markets/#{exchange}/#{market}/ohlc?after=#{unixtime}&periods=#{period}")

      response = Request.get(url)

      case response
      when Net::HTTPSuccess
        json = Utils.parse_json(response.body)
        raise JSONError.new(response) unless json.is_a?(Hash)

        data = json['result']
        allowance = json['allowance']
        raise JSONError.new(response) unless data.is_a?(Hash) && allowance.is_a?(Hash)

        timestamp, o, h, l, c, volume = best_matching_record(data, unixtime, current_time)
        raise NoDataError.new(response, 'No price data returned') unless timestamp && o

        actual_time = Time.at(timestamp)

        return DataPoint.new(
          price: o,
          time: actual_time,
          api_time_spent: allowance['cost'],
          api_time_remaining: allowance['remaining']
        )
      when Net::HTTPNotFound
        raise UnknownCoinError.new(response)
      when Net::HTTPClientError
        raise BadRequestError.new(response)
      else
        raise ServiceUnavailableError.new(response)
      end
    end


    private

    def best_matching_record(data, unixtime, request_time)
      candidates = []

      data.keys.sort_by { |k| k.to_i }.each do |k|
        records = data[k] || []
        previous = nil

        records.each do |record|
          timestamp, o, h, l, c, volume = record

          if timestamp >= unixtime
            candidates.push(record) unless timestamp > request_time
            break
          else
            previous = record
          end
        end

        candidates.push(previous) if previous
      end

      candidates.sort_by { |record| (record[0] - unixtime).abs }.first
    end

    def precision_for_time(time)
      now = Time.now

      DAYS_FOR_PERIODS.keys.sort.each do |period|
        days = DAYS_FOR_PERIODS[period]
        earliest_date = now - days * 86400

        if earliest_date < time
          return period
        end
      end

      nil
    end

    def get_exchanges
      url = URI("#{BASE_URL}/exchanges")

      response = Request.get(url)

      case response
      when Net::HTTPSuccess
        json = Utils.parse_json(response.body)
        raise JSONError.new(response) unless json.is_a?(Hash) && json['result'].is_a?(Array)

        return json['result'].select { |e| e['active'] == true }.map { |e| e['symbol'] }.sort
      when Net::HTTPClientError
        raise BadRequestError.new(response)
      else
        raise ServiceUnavailableError.new(response)
      end
    end
  end
end
